pkgs/crypto/crypto.go

```diff
@@ -587,3 +595,95 @@ func EncryptPrivateKey(priv []byte, keyStorePassword string) ([]byte, error) {
+	sharesExpectedLength := EncryptedKeyLength*len(reshare.OldOperators) + pubKeysOffset
+	if len(reshare.Keyshares) != sharesExpectedLength {
+		return nil, fmt.Errorf("GetPubCommitsFromSharesData: shares data len is not correct, expected %d, actual %d", sharesExpectedLength, len(reshare.Keyshares))
+	}
+	pubKeys := utils.SplitBytes(reshare.Keyshares[signatureOffset:pubKeysOffset], phase0.PublicKeyLength)
+	// try to recover commits
+	var kyberPubShares []*share.PubShare
+	for i, pubk := range pubKeys {
+		blsPub := &bls.PublicKey{}
+		err := blsPub.Deserialize(pubk)
+		if err != nil {
+			return nil, err
+		}
+		v := suite.G1().Point()
+		err = v.UnmarshalBinary(blsPub.Serialize())
+		if err != nil {
+			return nil, err
+		}
+		kyberPubhare := &share.PubShare{
+			I: int(i),
+			V: v,
+		}
+		kyberPubShares = append(kyberPubShares, kyberPubhare)
+	}
+	pubPoly, err := share.RecoverPubPoly(suite.G1(), kyberPubShares, int(reshare.OldT), len(reshare.OldOperators))
+	if err != nil {
+		return nil, err
+	}
+	_, commits := pubPoly.Info()
+	return commits, nil
+}
+
+func GetSecretShareFromSharesData(reshare *wire.Reshare, opPrivateKey *rsa.PrivateKey, operatorID uint64) (*share.PriShare, error) {
+	suite := kyber_bls12381.NewBLS12381Suite()
+	signatureOffset := phase0.SignatureLength
+	pubKeysOffset := phase0.PublicKeyLength*len(reshare.OldOperators) + signatureOffset
+	sharesExpectedLength := EncryptedKeyLength*len(reshare.OldOperators) + pubKeysOffset
+	if len(reshare.Keyshares) != sharesExpectedLength {
+		return nil, fmt.Errorf("GetSecretShareFromSharesData: shares data len is not correct, expected %d, actual %d", sharesExpectedLength, len(reshare.Keyshares))
+	}
+	encryptedKeys := utils.SplitBytes(reshare.Keyshares[pubKeysOffset:], len(reshare.Keyshares[pubKeysOffset:])/len(reshare.OldOperators))
+	// try to decrypt private share
+	var kyberPrivShare *share.PriShare
+	for _, enck := range encryptedKeys {
+		prShare, err := rsaencryption.DecodeKey(opPrivateKey, enck)
+		if err != nil {
+			continue
+		}
+		secret := &bls.SecretKey{}
+		err = secret.SetHexString(string(prShare))
+		if err != nil {
+			return nil, err
+		}
+		// Check operator signature
+		var initVerify bool
+		initiatorPubKey, err := ParseRSAPubkey(reshare.InitiatorPublicKey)
+		if err != nil {
+			return nil, err
+		}
+		encInitPub, err := EncodePublicKey(initiatorPubKey)
+		if err != nil {
+			return nil, err
+		}
+		sigs := utils.SplitBytes(reshare.CeremonySigs, SignatureLength)
+		for _, sig := range sigs {
+			dataToVerify := make([]byte, len(secret.Serialize())+len(encInitPub))
+			copy(dataToVerify[:len(secret.Serialize())], secret.Serialize())
+			copy(dataToVerify[len(secret.Serialize()):], encInitPub)
+			err := VerifyRSA(&opPrivateKey.PublicKey, dataToVerify, sig)
+			if err != nil {
+				continue
+			}
+			initVerify = true
+			break
+		}
+		if !initVerify {
+			return nil, fmt.Errorf("cant verify initiator public key")
+		}
```

Comment:
so you go through all the sigs but only one is enough to verify? shouldn't be opposite?

Reply:
Each operator verifies its own sig because we sign `operator`s secret + initiator pub` to have a relation between them . To verify for all I would need to hash it and send this hash in addition to sig, so more stuff to transfer

Reply:
it will probably be better to verify all operators in this ceremony, we might want to do it on another PR though. @moshe-blox , @MatusKysel wdyt?

Reply:
seems like the message we're verifying contains the share's BLS private key, so it would be impossible to verify other operator's signatures, though i'm not sure why that message couldn't contain the public key instead, care to explain? @pavelkrolevets 
*(i still have lots to catch up on with DKG)*

btw, is the order of signatures in the CeremonySigs slice deterministic (or provided elsewhere in the reshare message), so you can know at which index yours is? if so, we could verify just one.

regardless, we should move serializations/allocations out of the loop:
```go
dataToVerify := make([]byte, len(serialized)+len(encInitPub))
copy(dataToVerify[:len(serialized)], serialized)
copy(dataToVerify[len(serialized):], encInitPub)
for _, sig := range sigs {
	err := VerifyRSA(&opPrivateKey.PublicKey, dataToVerify, sig)
	if err != nil {
		continue
	}
	initVerify = true
	break
}
```

Reply:
@moshe-blox yeah true about serializations/allocations at loop, my bad. 
I was thinking about using public instead, but it needs to be redesigned. Ill create a ticket for it.